converters/toarray.ts

```diff
@@ -0,0 +1,82 @@
+import { IEvent, IEventArray, ILayoutState } from "../declarations/clarity";
+import StateManager from "./../src/plugins/layout/statemanager";
+import { getNodeIndex } from "./../src/plugins/layout/stateprovider";
+import { hashCode } from "./../src/utils";
+import schemas from "./schema";
+
+// We serialize send a lot of JSONs with identical structures and putting all property names on the
+// wire every time can be very redundant. We can save bytes by splitting JSON into a separate nested array of
+// its property values and a separate nested array of its property names (schema). This way, each unique schema
+// has to only be sent for the first event of that type and the following events can then re-use this schema
+// on the server for reconstructing the full JSON.
+export default function(event: IEvent): IEventArray {
+  let keys = Object.keys(event);
+  let schema = schemas.createSchema(event.data);
+  let newSchema = schemas.addSchema(schema);
+  let dataArray = dataToArray(event.data);
+  let schemaPayload = newSchema ? schema : schemas.getSchemaHashcode(schema);
+  let array = [event.id, event.origin, event.type, event.time, dataArray, schemaPayload] as IEventArray;
+  return array;
+}
+
+// When we discover the entire document or a subtree for the first time, we can structure our payload in a way
+// such that some information can be inferred from the structure, as opposed to being written explicitly and this
```

Comment:
>some information can be inferred from the structure [](start = 13, length = 51)

Let's not add this complexity since we are now dealing with redundant data pretty well in your code flow. This is an optimization we can revisit. I know this is where we started, but for the sake of simplicity, let's avoid this extra complexity.